Mathlib/RingTheory/Etale/Field.lean

```diff
@@ -148,4 +153,61 @@ theorem iff_isSeparable [EssFiniteType K L] :
     FormallyEtale K L ↔ Algebra.IsSeparable K L :=
   ⟨fun _ ↦ FormallyUnramified.isSeparable K L, fun _ ↦ of_isSeparable K L⟩
 
+attribute [local instance]
+  IsArtinianRing.subtype_isMaximal_finite IsArtinianRing.fieldOfSubtypeIsMaximal in
+/--
+If `A` is an essentially of finite type algebra over a field `K`, then `A` is formally etale
+over `K` if and only if `A` is a finite product of separable field extensions.
+-/
+theorem iff_exists_algEquiv_prod [EssFiniteType K A] :
+    FormallyEtale K A ↔
+      ∃ (I : Type u) (_ : Finite I) (Ai : I → Type u) (_ : ∀ i, Field (Ai i))
+        (_ : ∀ i, Algebra K (Ai i)) (_ : A ≃ₐ[K] Π i, Ai i),
+        ∀ i, Algebra.IsSeparable K (Ai i) := by
```

Comment:
For this type of theorems, I think it might be better to separate the forward and backward directions. For the forward direction you can use `Fin n` instead of an arbitrary indexing type. For the backward direction you can take `I` and `Ai` in any universe, and the conclusion could simply say `FormallyEtale K (Π i, Ai i)` (this could be an instance, but a more generaly instance would assume each `Ai i` is formally etale); you can add a separate lemma saying that `FormallyEtale K` is invariant under `K`-AlgHoms. This would ensure both directions are maximally general, but I understand it may need additional infrastructure, like `congrLeft` constructions (on the indexing type) for various kinds of algebraic isomorphisms (AlgHom here).